### PR TITLE
I140 make leaderboard page

### LINKED ISSUE #140: Make Leaderboard page
Make a page which displays and ranks the users who have completed the most mileage

### PR DESCRIPTION
**Pull request by HuxleyBerry**
## Change Summary

**[Briefly summarise the changes that you made. Just high-level stuff]**
* Added a total_mileage field to the user and team models
    * If you achieve some mileage and then join a team, the already achieved mileage does not go towards your team's total. However, once you join the team, all mileage you achieve from then on will be counted.
* Added signals to respond to any updating, creating or deleting mileage model instances by modifying the total_mileage field accordingly
    * This could have been done in the post_mileage view instead, but the advantage of using signals is that the total_mileage field will also be updated if the admin messes with users' mileage
* Added a view in the backend to calculate the user and team leaderboards (just the top 100). If a user or team is passed with the request, the individual ranking of the passed user/team is also calculated (even if they're not in the top 100)
* Made the leaderboard page on the frontend. It's pretty similar to the design on figma

### Change Form

**Fill this up (NA if not available). If a certain criteria is not met, can you please give a reason.**

- [x] The pull request title has an issue number
- [x] The change works by "Smoke testing" or quick testing
- [ ] The change has tests
- [ ] The change has documentation

## Other Information

**[Is there anything in particular in the review that I should be aware of?]**

# Related Issue

- Resolve #140

### CODE REVIEW

File: client/src/views/LeaderboardView.vue [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+        </td>
+        <td v-else class="text-right text-subtitle-1">{{ currentUser.rank }}</td>
+        <td class="text-subtitle-1">{{ currentUser.username }}</td>
+        <td>
+          <v-chip color="green" class="rounded text-h6 w-100 d-flex justify-center">{{
+            currentUser.totalMileage
+          }}</v-chip>
+          <!-- <p class="text-subtitle-2 text-center">TOTAL MILEAGE</p> -->
+        </td>
+      </tr>
```
Comment (ChantelleKerr) [inline client/src/views/LeaderboardView.vue:55]: I would try make the styling slightly different than the other rows

File: client/src/views/LeaderboardView.vue [hunk 2]
```diff
@@ -0,0 +1,205 @@
+<template>
+  <div class="header">
+    <h1 class="text-center text-secondaryTint font-weight-medium text-md-h1" id="title">
+      Leaderboards
+    </h1>
+  </div>
```
Comment (ChantelleKerr) [inline client/src/views/LeaderboardView.vue:6]: the colour looks different compared to other pages.

File: client/src/views/LeaderboardView.vue [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+      density="compact"
+      width="120px"
+      :color="activeButton === board ? 'rgb(52, 94, 158, 0.3)' : ''"
+      :class="activeButton === board ? 'text-secondaryBlue' : 'black'"
+      v-for="board in ['Individual', 'Teams']"
+      :key="board"
+      rounded="xl"
+      @click="activeButton = board"
+      >{{ board }}</v-btn
+    >
```
Comment (ChantelleKerr) [inline client/src/views/LeaderboardView.vue:19]: needs slightly more spacing top and bottom

File: server/api/users/migrations/0009_user_total_mileage.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+        ('users', '0008_alter_user_avatar'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='user',
+            name='total_mileage',
+            field=models.FloatField(default=0.0),
+        ),
+    ]
```
Comment (ChantelleKerr) [inline server/api/users/migrations/0009_user_total_mileage.py:18]: I would suggest deleting 0009 and 0011 migrations - merge main into this branch. and then makemigrations again. Main already has a 0009 and 0010.

From my experience on the wais project deleting migration files that has already been on deployed causes problems for production version.

**Review by ChantelleKerr — CHANGES_REQUESTED** [review]
Some changes - its looking very nice!
While your here can you please make the events and leaderboard banner the same height? I like the height of the leaderboard :)

**Comment by ChantelleKerr** [comment]
Might have to merge main into your branch. You could change how you get the total mileage since you have it stored in the user table

**Comment by HuxleyBerry** [comment]
> Might have to merge main into your branch. You could change how you get the total mileage since you have it stored in the user table

Hmm, so calling get_mileage() for every user rather than simply accessing the total_mileage field? The latter option should be a lot more efficient I think.

**Comment by ChantelleKerr** [comment]
> > Might have to merge main into your branch. You could change how you get the total mileage since you have it stored in the user table
> 
> Hmm, so calling get_mileage() for every user rather than simply accessing the total_mileage field? The latter option should be a lot more efficient I think.

Yeah I agree, the graph PR didnt have a total_mileage field so yours works well i think :)

**Comment by HuxleyBerry** [comment]
However, in the teams case, I realise there is the potential for there to be a discrepancy between the value of total_mileage in the team model, and the sum value obtained by get_mileage(). This is because total_mileage doesn't count mileage achieved before a user joined the team, whereas get_mileage() does. So for consistency, maybe get_mileage() should return the value of total mileage in this case. We wouldn't want the total km value in the team dashboard to be different to what is shown on the leaderboard. The downside to this is that the total implied by the team mileage graph wouldn't match with the total given in the team dashboard. Not sure what to best thing to do is...

**Comment by HuxleyBerry** [comment]
Perhaps the solution is for the mileage model to have a nullable team field, which records the team (if any) the user was in when the mileage was achieved. I think that will fix everything

...assuming that what we want to do is only count mileage achieved when the user is part of the team

**Comment by HuxleyBerry** [comment]
Think I'm done now

**Review by ChantelleKerr — APPROVED** [review]
Looks really nice! Good stuff :D

**Pull request merged**